common/workunit/workunit.cpp

```diff
@@ -6805,6 +6838,119 @@ const char *clusterTypeString(ClusterType clusterType, bool lcrSensitive)
+{
+    if (!isWorkunitDAToken(distributedAccessToken))
+    {
+        DBGLOG("Not a valid workunit distributed access token");
+        return false;
+    }
+
+    //Extract the string between [ and ]
+    const char * pEndBracket = strchr(distributedAccessToken + 5, ']');
+    StringBuffer tokenValues;
+    tokenValues.append(pEndBracket - distributedAccessToken - 5, distributedAccessToken + 5);
+
+    StringArray nameValues;
+    nameValues.appendList(tokenValues.str(), ",",true);
+
+    if (user || wuid)
+    {
+        for (int x = 0; x < nameValues.ordinality(); x++)
+        {
+            if (user && 0==strncmp(nameValues[x],"u=",2))//Extract user
+                user->append(nameValues[x] + 2);
+            if (wuid && 0==strncmp(nameValues[x],"w=",2))//Extract wuid
+                wuid->append(nameValues[x] + 2);
+        }
+    }
+
+    if (privKey)
+    {
+        const char * finger = pEndBracket;
+        ++finger;
+        while (*finger && *finger != ';')
+            privKey->append(1, finger++);
+    }
+    return true;
+}
+
+//Verifies the given signed workunit distributed access token was created
+//and signed by the given wuid and user, and that the workunit is still active
+//Returns:
+//   0 : Success, token is valid and workunit is active
+//   1 : Signature does not verify (wuid/username don't match, or signature does not verify)
+//   2 : Workunit not active
+//   Throws if unable to open workunit
+wuTokenStates verifyWorkunitDAToken(const char * distributedAccessToken)
+{
+    StringBuffer tokWuid;
+    StringBuffer tokUser;
+    if (!extractFromWorkunitDAToken(distributedAccessToken, &tokWuid, &tokUser, nullptr))//get the wuid and user
+    {
+        //Not a valid workunit distributed access token
+        return wuTokenInvalid;
+    }
+
+    //Validate signature
+    IDigitalSignatureManager * pDSM = queryDigitalSignatureManagerInstanceFromEnv();
+    if (pDSM && pDSM->isDigiVerifierConfigured())
+    {
+        const char * finger;
+        StringBuffer token;//receives copy of everything up until signature
+        for (finger = distributedAccessToken; *finger && *finger != ';'; finger++)
+            token.append(1, finger);
+        token.append(1, finger);//append ;
+
+        StringBuffer sig(++finger);
+        if (!pDSM->digiVerify(sig, token))
+        {
+            ERRLOG("verifyWorkunitDAToken : workunit distributed access token does not verify");
+            return wuTokenInvalid;
+        }
+    }
+
+    Owned<IWorkUnitFactory> factory = getWorkUnitFactory();
+    Owned<IConstWorkUnit> cw = factory->openWorkUnit(tokWuid.str());
+    if(!cw)
+    {
+        throw MakeStringException(WUERR_WorkunitAccessDenied,"verifyWorkunitDAToken : Cannot open workunit %s",tokWuid.str());
+    }
+
+    //Verify user matches
+    if (!isEmptyString(cw->queryUser()) || !isEmptyString(tokUser.str()))
+    {
+        if (!streq(cw->queryUser(), tokUser.str()))
```

Comment:
need to be careful - cw->queryUser() could be null - although someone would need to be deliberately trying to break the system.  E.g. create a workunit on a system with no security, and then send in a request using a token that does include a user.